read_hbtplus.py

```diff
@@ -0,0 +1,354 @@
+            nr_files = int(infile["NumberOfFiles"][...])
+    else:
+        nr_files = None
+    nr_files = comm.bcast(nr_files)
+
+    # Assign files to MPI ranks
+    files_per_rank = np.zeros(comm_size, dtype=int)
+    files_per_rank[:] = nr_files // comm_size
+    files_per_rank[: nr_files % comm_size] += 1
+    assert np.sum(files_per_rank) == nr_files
+    first_file_on_rank = np.cumsum(files_per_rank) - files_per_rank
+
+    # Read in the halos from the HBT output:
+    # 'halos' will be an array of structs with the halo catalogue
+    # 'ids_bound' will be an array of particle IDs in halos, sorted by halo
+    halos = []
+    ids_bound = []
+    for file_nr in range(
+        first_file_on_rank[comm_rank],
+        first_file_on_rank[comm_rank] + files_per_rank[comm_rank],
+    ):
+        with h5py.File(hbt_filename(basename, file_nr), "r") as infile:
+            halos.append(infile["Subhalos"][...])
+            ids_bound.append(infile["SubhaloParticles"][...])
+
+    # Get the dtype for particle IDs
+    if len(ids_bound) > 0:
+        id_dtype = h5py.check_vlen_dtype(ids_bound[0].dtype)
+    else:
+        id_dtype = None
+
+    # Concatenate arrays of halos from different files
+    if len(halos) > 0:
+        halos = np.concatenate(halos)
+    else:
+        # This rank was assigned no files
+        halos = None
+    halos = virgo.mpi.util.replace_none_with_zero_size(halos, comm=comm)
+
+    # Combine arrays of particles in halos
+    if len(ids_bound) > 0:
+        ids_bound = np.concatenate(ids_bound)  # Combine arrays of halos from different files
+        if len(ids_bound) > 0:
+            ids_bound = np.concatenate(ids_bound)  # Combine arrays of particles from different halos
+        else:
+            # The files assigned to this rank contain zero halos
+            ids_bound = np.zeros(0, dtype=id_dtype)
+    else:
+        # This rank was assigned no files
+        ids_bound = None
+    ids_bound = virgo.mpi.util.replace_none_with_zero_size(ids_bound, comm=comm)
+
+    # Assign halo indexes to the particles
+    nr_local_halos = len(halos)
+    total_nr_halos = comm.allreduce(nr_local_halos)
+    halo_offset = comm.scan(len(halos), op=MPI.SUM) - len(halos)
+    halo_index = np.arange(nr_local_halos, dtype=int) + halo_offset
+    halo_size = halos["Nbound"]
+    del halos
+    grnr_bound = np.repeat(halo_index, halo_size)
+
+    # Assign ranking by binding energy to the particles
+    rank_bound = -np.ones(grnr_bound.shape[0], dtype=int)
+    offset = 0
+    for halo_nr in range(nr_local_halos):
+        rank_bound[offset : offset + halo_size[halo_nr]] = np.arange(
+            halo_size[halo_nr], dtype=int
+        )
+        offset += halo_size[halo_nr]
+    assert np.all(rank_bound >= 0)  # HBT only outputs bound particles
+    del halo_size
+    del halo_offset
+    del halo_index
+
+    # Now check for duplicates. HBTplus can assign the same particle to multiple
```

Comment:
I would be inclined to convert it to a check, just in case we encounter any duplicates in the 10K run from a mechanism we haven't thought of yet. 

Should we get rid of the halo sizes file now we don't have duplicates?

Reply:
I think so. We should be able to just use Nbound from the HBT output now.

Reply:
I've removed the halo sizes in https://github.com/SWIFTSIM/SOAP/pull/71/commits/a113035145cc655af6a8465a43895218023d6276

Reply:
I've stripped out the code that assigned duplicate particles to a single subhalo, and added an assert to check we don't have duplicates in https://github.com/SWIFTSIM/SOAP/pull/71/commits/019da21267f978bdac47c8b3f106b22ba303b68c